### ISSUE TITLE
sources_default not being consistently honored, nor is IRRd style '!s'

### ISSUE THREAD

**Issue by ccaputo** [irrdnet/irrd#635]
**Describe the bug**
I upgraded from 4.1.7 to 4.2.2 today and noticed query results have changed substantially.

Digging in, it appears that sources_default is not working for IRRd style queries, nor is IRRd style '!s'.

Given:

```
    sources_default:
        - NTTCOM
        - RADB
        - RIPE
        - RIPE-NONAUTH
        - ALTDB
        - BELL
        - LEVEL3
        - RGNET
        - APNIC
        - JPIRR
        - ARIN
        - ARIN-NONAUTH
        - BBOI
        - TC
        - AFRINIC
        - CANARIE
        - ARIN-WHOIS
        - NTT-RPKI
        - REGISTROBR
```

and

```
as-set:         AS-RS
descr:          TC Radio Systems Clients
members:        AS20539
members:        AS13202
members:        AS16327
members:        AS60313
members:        AS48892
members:        AS-CHEREDA-SM
members:        AS50843
tech-c:         DUMY-RIPE
admin-c:        DUMY-RIPE
mnt-by:         TCRS-MNT
created:        2009-10-29T13:23:14Z
last-modified:  2020-09-11T09:45:42Z
source:         RIPE
remarks:        ****************************
remarks:        * THIS OBJECT IS MODIFIED
remarks:        * Please note that all data that is generally regarded as personal
remarks:        * data has been removed from this object.
remarks:        * To view the original object, please query the RIPE Database at:
remarks:        * http://www.ripe.net/whois
remarks:        ****************************

as-set:         AS-RS
descr:          Provides transit to this motley cast of characters
members:        AS3066, AS3944, AS7247, AS7432, AS11893,
                       AS16668, AS19363, AS21659, AS22779
admin-c:        RES70
tech-c:         RES70
notify:         rs@seastrom.com
mnt-by:         MAINT-AS3066-ALTDB
changed:        rs@seastrom.com 20040116
source:         ALTDB
```

Some example results:

```
!iAS-RS,1
A68
AS11893 AS16668 AS19363 AS21659 AS22779 AS3066 AS3944 AS7247 AS7432
C
```

The above is the ALTDB result, rather than the expected RIPE result based on `sources_default` above.

```
!!
!sNTTCOM,RADB,RIPE,RIPE-NONAUTH,ALTDB,BELL,LEVEL3,RGNET,APNIC,JPIRR,ARIN,ARIN-NONAUTH,BBOI,TC,AFRINIC,CANARIE,ARIN-WHOIS,NTT-RPKI,REGISTROBR
!iAS-RS,1
C
A68
AS11893 AS16668 AS19363 AS21659 AS22779 AS3066 AS3944 AS7247 AS7432
C
!sRADB,RIPE,RIPE-NONAUTH,ALTDB,BELL,LEVEL3,RGNET,APNIC,JPIRR,ARIN,ARIN-NONAUTH,BBOI,TC,AFRINIC,CANARIE,ARIN-WHOIS,NTT-RPKI,REGISTROBR
C
!iAS-RS,1
A97
AS13202 AS16327 AS197058 AS20539 AS21312 AS24999 AS25505 AS42530 AS48892 AS50843 AS51874 AS60313
C
```

In the above, the removal of `NTTCOM` changed the result from being from ALTDB to being from RIPE. The removal of `NTTCOM` should not have had an impact on the above, nor was the initial result correct given that RIPE was before ALTDB in the argument list. I am not certain that IRRd style `!s` implies priority, but it seems to:

```
!!
!sRIPE,ALTDB
C
!iAS-RS,1
A97
AS13202 AS16327 AS197058 AS20539 AS21312 AS24999 AS25505 AS42530 AS48892 AS50843 AS51874 AS60313
C
!sALTDB,RIPE
C
!iAS-RS,1
A68
AS11893 AS16668 AS19363 AS21659 AS22779 AS3066 AS3944 AS7247 AS7432
C
```

Regardless, the removal of `NTTCOM` above should not have changed the results.

So there seem to be bugs with respect to honoring `sources_default` and with the evaluation of `!s`.

**IRRd version you are running**
4.2.2

**Comment by mxsasha** [comment]
This is very curious, because I can't reproduce it with v4.2.2. It works exactly as expected for me (my sources_default also places RIPE before ALTDB):

```
!!
AS-RS
as-set:         AS-RS
descr:          Provides transit to this motley cast of characters
members:        AS3066, AS3944, AS7247, AS7432, AS11893,
	               AS16668, AS19363, AS21659, AS22779
admin-c:        RES70
tech-c:         RES70
notify:         rs@seastrom.com
mnt-by:         MAINT-AS3066-ALTDB
changed:        rs@seastrom.com 20040116
source:         ALTDB

as-set:         AS-RS
descr:          TC Radio Systems Clients
members:        AS20539
members:        AS13202
members:        AS16327
members:        AS60313
members:        AS48892
members:        AS-CHEREDA-SM
members:        AS50843
tech-c:         DUMY-RIPE
admin-c:        DUMY-RIPE
mnt-by:         TCRS-MNT
created:        2009-10-29T13:23:14Z
last-modified:  2020-09-11T09:45:42Z
source:         RIPE
remarks:        ****************************
remarks:        * THIS OBJECT IS MODIFIED
remarks:        * Please note that all data that is generally regarded as personal
remarks:        * data has been removed from this object.
remarks:        * To view the original object, please query the RIPE Database at:
remarks:        * http://www.ripe.net/whois
remarks:        ****************************


!iAS-RS
A62
AS-CHEREDA-SM AS13202 AS16327 AS20539 AS48892 AS50843 AS60313
C
!sALTDB,RIPE
C
!iAS-RS
A68
AS11893 AS16668 AS19363 AS21659 AS22779 AS3066 AS3944 AS7247 AS7432
C
!sRIPE,ALTDB
C
!iAS-RS
A62
AS-CHEREDA-SM AS13202 AS16327 AS20539 AS48892 AS50843 AS60313
C
!iAS-RS,1
A97
AS13202 AS16327 AS197058 AS20539 AS21312 AS24999 AS25505 AS42530 AS48892 AS50843 AS51874 AS60313
C
```

You are correct that `!s` implies priority, and that removal or addition of NTTCOM should not matter for this query.

**Comment by ccaputo** [comment]
> This is very curious, because I can't reproduce it with v4.2.2. It works exactly as expected for me (my sources_default also places RIPE before ALTDB):

Additionally, the bug comes and goes. Sometimes I get the correct result, other times not. For example:

```
$ echo -e "\x21iAS-RS,1" | nc s 43
A97                      
AS13202 AS16327 AS197058 AS20539 AS21312 AS24999 AS25505 AS42530 AS48892 AS50843 AS51874 AS60313
C                                                                                        
$ echo -e "\x21iAS-RS,1" | nc s 43
A97                  
AS13202 AS16327 AS197058 AS20539 AS21312 AS24999 AS25505 AS42530 AS48892 AS50843 AS51874 AS60313
C                                 
```

and then minutes later without a restart of anything:

```
$ echo -e "\x21iAS-RS,1" | nc s 43
A68
AS11893 AS16668 AS19363 AS21659 AS22779 AS3066 AS3944 AS7247 AS7432
C
$ echo -e "\x21iAS-RS,1" | nc s 43
A68
AS11893 AS16668 AS19363 AS21659 AS22779 AS3066 AS3944 AS7247 AS7432
C
```

**Comment by ccaputo** [comment]
In case helpful:

```
!!
!iAS-RS,1
A68
AS11893 AS16668 AS19363 AS21659 AS22779 AS3066 AS3944 AS7247 AS7432
C
!sNTTCOM,RADB,RIPE,RIPE-NONAUTH,ALTDB,BELL,LEVEL3,RGNET,APNIC,JPIRR,ARIN,ARIN-NONAUTH,BBOI,TC,AFRINIC,CANARIE,ARIN-WHOIS,NTT-RPKI,REGISTROBR                                      
C
!iAS-RS,1
A68
AS11893 AS16668 AS19363 AS21659 AS22779 AS3066 AS3944 AS7247 AS7432
C
!sRADB,RIPE,RIPE-NONAUTH,ALTDB,BELL,LEVEL3,RGNET,APNIC,JPIRR,ARIN,ARIN-NONAUTH,BBOI,TC,AFRINIC,CANARIE,ARIN-WHOIS,NTT-RPKI,REGISTROBR
C
!iAS-RS,1
A97
AS13202 AS16327 AS197058 AS20539 AS21312 AS24999 AS25505 AS42530 AS48892 AS50843 AS51874 AS60313
C
!sNTTCOM,RADB,RIPE,RIPE-NONAUTH,ALTDB,BELL,LEVEL3,RGNET,APNIC,JPIRR,ARIN,ARIN-NONAUTH,BBOI,TC,AFRINIC,CANARIE,ARIN-WHOIS,NTT-RPKI,REGISTROBR                                      
C
!iAS-RS,1
A68
AS11893 AS16668 AS19363 AS21659 AS22779 AS3066 AS3944 AS7247 AS7432
C
```

Am seeing that when I include NTTCOM in the list above, I am matching sources_default. This alters the SQL query to not include mention of each source.

I suspect the non-determinism above is due to the whims of PostgreSQL.

I have submitted a possible solution as a pull request #637. The solution has made things be deterministic, and correct in our use.

**Comment by mxsasha** [comment]
Nice find, it's merged, thanks!

**Comment by mxsasha** [comment]
Released in 4.2.4
